### PR TITLE
Changes to ViscoPlasticityStressUpdateBase and SingleVariableReturnMappingSolution to help downstream development

### PR DESCRIPTION
**Pull request by tophmatthews**
Add members to ViscoPlasticityStressUpdateBase to make downstream models easier to create.
Combined ADSingleVariableReturnMappingSolution and SingleVariableReturnMappingSolution.

Ref #20290

### CODE REVIEW

**Comment by tophmatthews** [comment]
Actually, it should be relatively easy to combine `ADSingleVariableReturnMappingSolution` and `SingleVariableReturnMappingSolution` if there was a `GenericChainedReal`

**Comment by dschwen** [comment]
> Actually, it should be relatively easy to combine `ADSingleVariableReturnMappingSolution` and `SingleVariableReturnMappingSolution` if there was a `GenericChainedReal`

There is. It is called `VariableAndDerivative<is_ad>` and I'm adding that in my PR.

**Comment by tophmatthews** [comment]
> Oh wait, no. You mean something that switches between `ChainedReal` and `ChainedADReal`?

Yeah, I'm doing it here with
```
template <bool is_ad>
using GenericChainedReal = typename std::conditional<is_ad, ChainedADReal, ChainedReal>::type;
```
There might be a better way, but this works just fine here.

**Comment by tophmatthews** [comment]
Ready for your 👀  @dschwen

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h [hunk 1]
```diff
@@ -32,23 +40,25 @@ class SingleVariableReturnMappingSolution
...
+  minimumPermissibleValue(const GenericReal<is_ad> & effective_trial_stress) const;
 
   /**
    * Compute the maximum permissible value of the scalar.  For some models, the magnitude
    * of this may be known.
    * @param effective_trial_stress Effective trial stress
    */
-  virtual Real maximumPermissibleValue(const Real & effective_trial_stress) const;
+  virtual GenericReal<is_ad>
+  maximumPermissibleValue(const GenericReal<is_ad> & effective_trial_stress) const;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:61]: should those min/max permissible values really be dual numbers in the AD case?

Comment (dschwen) [reply modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:61]: I suppose it won;t hurt

**Review by dschwen — DISMISSED** [review]
Looks fine to me. Thanks.

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h (hunk 1 above)
Comment (tophmatthews) [reply modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:61]: Yes, I pass a ADReal in bison for this, based on other ADReals.

**Review by dschwen — DISMISSED** [review]

File: modules/tensor_mechanics/test/include/materials/PowerLawCreepTest.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "PowerLawCreepStressUpdate.h"
+#include "ADPowerLawCreepStressUpdate.h"
+
+template <bool is_ad>
+using PowerLawCreepStressUpdateTempl =
+    typename std::conditional<is_ad, ADPowerLawCreepStressUpdate, PowerLawCreepStressUpdate>::type;
```
Comment (tophmatthews) [inline modules/tensor_mechanics/test/include/materials/PowerLawCreepTest.h:17]: I have a follow on PR that will remove this, but I didn't want to pollute this here.

**Comment by tophmatthews** [comment]
If this doesn't boost coverage to 110% then I'm headed for the hills!

**Comment by dschwen** [comment]
```
tensor_mechanics/test:power_law_creep.bounds/high: Floating point exception signaled (invalid floating point operation)!
```

**Comment by tophmatthews** [comment]
Yeah...I'm not positive why that's happening though...almost seems like a std::pow with ADReal problem

**Comment by dschwen** [comment]
There is a real test failure that remains to be addressed.

**Comment by tophmatthews** [comment]
> There is a real test failure that remains to be addressed.

Yeah, this is stuck with the std::pow failure discussed on slack. I'm stumped on how to fix...

`stress_delta = (-1.15385e+13,{(0,-8.02536e+10), (1,-1.14641e+10), (2,3.66428e+10), (3,4.05387e+10), (4,2.15039e+10), (5,3.07179e+09), (6,2.21069e+10), (7,-3.21464e+10)})`

`const auto asdf2 = std::pow(MetaPhysicL::raw_value(stress_delta), 1.0);`<-- succeeds
`const auto asdf3 = std::pow(stress_delta, 1.0);`<-- fails, `EXC_ARITHMETIC`

**Comment by GiudGiud** [comment]
How does  libMesh::Utility::pow perform for this one?

**Comment by tophmatthews** [comment]
> How does libMesh::Utility::pow perform for this one?

Probably will work, but the non-int option needs to be retained in `ADPowerLawCreepStressUpdate`

**Comment by tophmatthews** [comment]
Skipping poorly performing test until https://github.com/roystgnr/MetaPhysicL/issues/82 is fixed

**Comment by tophmatthews** [comment]
Okay, I think it's ready for a new review @dschwen

**Comment by tophmatthews** [comment]
@dschwen  I'm still assuming that it's okay to skip the bad test even though https://github.com/libMesh/MetaPhysicL/pull/22 is in (e.g. wait until the next libmesh update?)

**Comment by tophmatthews** [comment]
ping @dschwen ?

**Review by dschwen — APPROVED** [review]

**Pull request merged**